doofinder.php

```diff
@@ -1690,6 +1747,149 @@ public function hookProductSearchProvider($params)
+    {
+        if (Configuration::get("DF_UPDATE_ON_SAVE")) {
+            $id_shop = $this->context->shop->id;
+            $this->addProductQueue($params["id_product"], $id_shop, "delete");
+
+            if($this->allowProcessProductsQueue()){
+                $this->processProductQueue($id_shop);
+            }
+        }
+    }
+
+    public function allowProcessProductsQueue(){
+        if (Configuration::get("DF_UPDATE_ON_SAVE")) {
+            $last_exec = Configuration::get("DF_UPDATE_ON_SAVE_LAST_EXEC", null, null, null, 0);
+            $delay = Configuration::get("DF_UPDATE_ON_SAVE_DELAY", null, null, null, 30);
+
+            if(is_int($delay)){
+                $last_exec_ts = strtotime($last_exec);
+
+                $diff_min = (time() - $last_exec_ts) / 60;
+                
+                if($diff_min > $delay){
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    public function setExecUpdateOnSave(){
+        Configuration::updateValue("DF_UPDATE_ON_SAVE_LAST_EXEC", date('Y-m-d H:i:s'));
+    }
+
+    public function addProductQueue($id_product, $id_shop, $action)
+    {
+        Db::getInstance()->insert(
+            "doofinder_product",
+            array(
+                "id_shop" => $id_shop,
+                "id_product" => $id_product,
+                "action" => $action,
+                "date_upd" => date('Y-m-d H:i:s')
+            ),
+            false,
+            true,
+            Db::REPLACE
+        );
+    }
+
+    public function processProductQueue($id_shop)
+    {   
+        $this->setExecUpdateOnSave();
+
+        $products_update = $this->getProductsQueue($id_shop);
+        $products_delete = $this->getProductsQueue($id_shop, "delete");
+
+        $languages = Language::getLanguages(true, $id_shop);
+        $currencies = Currency::getCurrenciesByIdShop($id_shop);
+
+        foreach ($languages as $language) {
+            foreach ($currencies as $currency) {
+                $this->sendProductsApi($products_update, $id_shop, $language['id_lang'], $currency['id_currency']);
+                $this->sendProductsApi($products_delete, $id_shop, $language['id_lang'], $currency['id_currency'], "delete");
+            }
+        }
+
+        $this->deleteProductQueue($id_shop);
+    }
+
+    public function getProductsQueue($id_shop, $action = "update")
+    {
+        $products = Db::getInstance()->executeS(
+            "
+            SELECT id_product FROM " . _DB_PREFIX_ . "doofinder_product 
+            WHERE action = '" . pSQL($action) . "' AND id_shop = " . (int)$id_shop
+        );
+
+        return array_column($products, "id_product");
+    }
+
+    public function deleteProductQueue($id_shop){
+        Db::getInstance()->execute("DELETE from " . _DB_PREFIX_ . "doofinder_product WHERE id_shop = " . (int)$id_shop);
+    }
+
+    public function sendProductsApi($products, $id_shop, $id_lang, $id_currency, $action = "update")
```

Comment:
Two explicit functions, one for update and the other for delete and a third functions with the common code to be called in update & delete?

Reply:
The **hookActionProductSave** and **hookActionProductDelete** functions are the ones triggered by Prestashop when the corresponding modification is made to the product. This is managed by Prestashop like this.

Then a function is called to add to the queue, which is common for update/remove.
And then a function is called to process the queue, which processes both cases (modified and deleted).